### PR TITLE
Update planning routers

### PR DESCRIPTION
**Pull request by duanxianbo**
[(task:20537)](https://staging.teamyapp.com/teams/107/tasks/20537)

### CODE REVIEW

File: src/libs/router/Route.tsx [hunk 1]
```diff
@@ -54,8 +55,10 @@ export class Route extends Component<Props, State> {
                     return;
                 }
 
+                const matchFunction = this.props.exact ? exactMatchWith : startsWith;
```
Comment (naturetime10) [inline src/libs/router/Route.tsx]:
```suggestion
                const matcher = this.props.exact ? exactMatchWith : startsWith;
```

File: src/libs/router/Route.tsx [hunk 2]
```diff
@@ -54,8 +55,10 @@ export class Route extends Component<Props, State> {
...
+
```
Comment (naturetime10) [inline src/libs/router/Route.tsx]:
```suggestion
```

File: src/libs/router/Route.tsx [hunk 3]
```diff
@@ -54,8 +55,10 @@ export class Route extends Component<Props, State> {
...
                 if (
-                    !startsWith(
+                    !matchFunction(
```
Comment (naturetime10) [inline src/libs/router/Route.tsx]:
```suggestion
                    !matcher(
```

File: src/libs/router/Route.tsx [hunk 4]
```diff
@@ -107,3 +110,17 @@ function startsWith(long: string[], prefix: string[]): boolean {
...
+    }
+
+    for (let index = 0; index < prefix.length; index++) {
+        if (long[index] !== prefix[index]) {
+            return false;
+        }
+    }
+
+    return true;
+}
```
Comment (naturetime10) [inline src/libs/router/Route.tsx]:
```suggestion
```

File: src/libs/router/Route.tsx (hunk 1 above)
Comment (naturetime10) [reply src/libs/router/Route.tsx]: Discussed offline. Let's move `src/routing/match.ts` into `src/libs/router` and use `exactMatch` from `match.ts`. Also let's string's `startWith` as you proposed!

File: src/routing/interceptors.ts [hunk 5]
```diff
@@ -293,6 +295,31 @@ const sprintsRouteInterceptor = (
...
+        if (!includePrefix(planningRoutes, routeUpdate.routePattern)) {
+            redirectToPlanningFallback(router, localStore.getState().currTeamId!);
+            return {
+                isRedirected: true
+            }
+        }
+
+        return {};
+    };
+
```
Comment (naturetime10) [inline src/routing/interceptors.ts]:
```suggestion
```

File: src/routing/interceptors.ts [hunk 6]
```diff
@@ -293,6 +295,31 @@ const sprintsRouteInterceptor = (
     };
 };
 
+
```
Comment (naturetime10) [inline src/routing/interceptors.ts]:
```suggestion
```

File: src/routing/interceptors.ts [hunk 7]
```diff
@@ -293,6 +295,31 @@ const sprintsRouteInterceptor = (
...
+            return {
+                isRedirected: true
+            }
+        }
+
+        return {};
+    };
+
+}
+
```
Comment (naturetime10) [inline src/routing/interceptors.ts]:
```suggestion
```

File: src/routing/routes.ts [hunk 8]
```diff
@@ -35,6 +41,13 @@ export const sprintRoutePattern = `${sprintsRoutePattern}/{${sprintIdParam}}`;
 export const sprintTabRoutePattern = `${sprintRoutePattern}/{${tabIdParam}}`;
 export const insightsRoutePattern = `${teamRoutePattern}/insights`;
 export const planningRoutePattern = `${teamRoutePattern}/planning`;
+
+export const projectsRoutePattern = `${planningRoutePattern}/projects`;
+export const projectRoutePattern = `${projectsRoutePattern}/{${projectIdParam}}`;
+export const phaseRoutePattern = `${projectRoutePattern}/phases/{${phaseIdParam}}`;
+export const storyRoutePattern = `${phaseRoutePattern}/stories/{${storyIdParam}}`;
+export const standaloneStoryRoutePattern = `${projectRoutePattern}/stories/{${storyIdParam}}`;
```
Comment (naturetime10) [inline src/routing/routes.ts]:
```suggestion
export const projectStoryRoutePattern = `${projectRoutePattern}/stories/{${storyIdParam}}`;
```

File: src/routing/routes.ts [hunk 9]
```diff
@@ -35,6 +41,13 @@ export const sprintRoutePattern = `${sprintsRoutePattern}/{${sprintIdParam}}`;
 export const sprintTabRoutePattern = `${sprintRoutePattern}/{${tabIdParam}}`;
 export const insightsRoutePattern = `${teamRoutePattern}/insights`;
 export const planningRoutePattern = `${teamRoutePattern}/planning`;
+
+export const projectsRoutePattern = `${planningRoutePattern}/projects`;
+export const projectRoutePattern = `${projectsRoutePattern}/{${projectIdParam}}`;
+export const phaseRoutePattern = `${projectRoutePattern}/phases/{${phaseIdParam}}`;
+export const storyRoutePattern = `${phaseRoutePattern}/stories/{${storyIdParam}}`;
```
Comment (naturetime10) [inline src/routing/routes.ts]:
```suggestion
export const phaseStoryRoutePattern = `${phaseRoutePattern}/stories/{${storyIdParam}}`;
```

File: src/routing/routes.ts [hunk 10]
```diff
@@ -60,6 +73,23 @@ export const taskPath = (teamId: number, taskId: number) =>
     `${teamPath(teamId)}/tasks/${taskId}`;
 export const insightsPath = (teamId: number) => `/teams/${teamId}/insights`;
 export const planningPath = (teamId: number) => `/teams/${teamId}/planning`;
+export const projectsPath = (teamId: number) => `/teams/${teamId}/planning/projects`;
+export const projectPath = (teamId: number, projectId: number) =>
+    `/teams/${teamId}/planning/projects/${projectId}`;
+export const phasePath = (teamId: number, projectId: number, phaseId: number) =>
+    `/teams/${teamId}/planning/projects/${projectId}/phases/${phaseId}`;
+export const storyPath = (
```
Comment (naturetime10) [inline src/routing/routes.ts]:
```suggestion
export const phaseStoryPath = (
```

File: src/routing/routes.ts [hunk 11]
```diff
@@ -60,6 +73,23 @@ export const taskPath = (teamId: number, taskId: number) =>
...
+    teamId: number,
+    projectId: number,
+    phaseId: number,
+    storyId: number,
+) =>
+    `/teams/${teamId}/planning/projects/${projectId}/phases/${phaseId}/stories/${storyId}`;
+export const standaloneStoryPath = (
```
Comment (naturetime10) [inline src/routing/routes.ts]:
```suggestion
export const projectStoryPath = (
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/libs/router/Route.tsx [hunk 12]
```diff
@@ -55,6 +57,7 @@ export class Route extends Component<Props, State> {
                 }
 
                 if (
+                    (this.props.exact && !exactMatch(route, routePattern)) ||
```
Comment (naturetime10) [inline src/libs/router/Route.tsx]: Can we split this check into multiple if statenment to improve readability.

File: src/libs/router/Route.tsx [hunk 13]
```diff
@@ -2,6 +2,7 @@ import React, { Component, ReactNode } from 'react';
 
 import { PopChannel } from '@lib/csp/csp';
 import { closeIfNot } from '@lib/csp/lib';
+import { exactMatch } from '@core/routing/match';
```
Comment (naturetime10) [inline src/libs/router/Route.tsx]: Can we move match into `@lib/router`?

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/libs/router/Route.tsx [hunk 14]
```diff
@@ -65,6 +74,7 @@ export class Route extends Component<Props, State> {
                     });
                     continue;
                 }
+
```
Comment (duanxianbo) [inline src/libs/router/Route.tsx:77]:
```suggestion
```

File: src/libs/router/router.ts [hunk 15]
```diff
@@ -349,3 +349,19 @@ function withDefaults(options?: NavigateOptions): NavigateOptions {
...
+export function includePrefix(patterns: string[], input: string): boolean {
+    for (const pattern of patterns) {
+        if (input.startsWith(pattern)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
```
Comment (naturetime10) [inline src/libs/router/router.ts]:
```suggestion
```

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**